man/rmf.man
.TH RMF %manext1% 1999-04-30 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
rmf \- remove an nmh folder
.SH SYNOPSIS
.HP 5
.na
.B rmf
.RB [ \-help ]
.RB [ \-version ]
.RI [ +folder ]
.RB [ \-interactive " | " \-nointeractive ]
.ad
.SH DESCRIPTION
.B rmf
removes all of the messages (files) within the specified (or default)
folder, and then removes the folder (directory) itself.
.PP
If there are any files within the folder which are not a part of
.BR nmh ,
they will
.I not
be removed, and an error will be produced.
If the folder is given explicitly, or the
.B \-nointeractive
option is given, then the folder will be removed without confirmation.
Otherwise,
the user will be asked for confirmation.  If
.B rmf
can't, for some reason, find the current folder the folder to be removed defaults
to `+inbox' (unless overridden by the user's profile entry \*(lqInbox\*(rq),
with confirmation.
.PP
If the folder being removed is a subfolder, the parent folder will become
the new current folder, and
.B rmf
will produce a message telling the user this has happened.
This provides an easy mechanism for selecting a set of messages,
operating on the list, then removing the list and returning to the
current folder, from which the list was extracted.
.PP
If
.B rmf
is used on a read-only folder, it will delete all the private sequences
(i.e., .RI \*(lqatr\- seq \- folder \*(rq entries) for this folder
from your context without affecting the folder itself.
.PP
.B rmf
irreversibly deletes messages that don't have other links, so
use it with caution.
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Current\-Folder:~^To find the default current folder
^Inbox:~^To find the default inbox
.fi
.SH "SEE ALSO"
.IR folder (1),
.IR rmm (1)
.SH DEFAULTS
.nf
.RB ` +folder "' defaults to the current folder, usually with confirmation"
.RB ` \-interactive "' if +folder' not given, `\-nointeractive' otherwise"
.fi
.SH CONTEXT
.B rmf
will set the current folder to the parent folder if a subfolder is removed;
or if the current folder is removed, it will make \*(lqinbox\*(rq current.
Otherwise, it doesn't change the current folder or message.
.SH BUGS
Although intuitively one would suspect that
.B rmf
works recursively, it does not.
Hence, if you have a sub-folder within a folder, in order to
.B rmf
the parent, you must first
.B rmf
each of the children.


NEWS
#
# NEWS - Release notes for nmh 1.6
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been nearly two years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

------------
NEW FEATURES
------------

The following are new features for the 1.6 release of nmh:

- All nmh programs will now warn about multiple profile entries for
  the same non-null, non-comment component.  In previous versions, all
  but the first were silently ignored.
- mhmail now supports all post(8) options, or optionally can use
  send(1) with all of its options.  Its formerly undocumented -resent
  switch has been documented.  And it has new -attach, -headerfield,
  and -send/-nosend switches.
- Default to flock() locking on OpenBSD and Darwin.
- Added -directives support to mhbuild(1) to control whether or not
  mhbuild will honor MIME directives by default.  And added support
  for special #on/#off/#pop directives to control the MIME directive
  processing state.
- Added -messageid switch to send(1) and post(8).  This allows
  selection of the style to use for generated Message-ID and
  Content-ID header fields.  The default localname style is
  pid.time@localname, where time is in seconds, and matches previous
  behavior.  The random style replaces the localname with some
  (pseudo)random bytes and uses microsecond-resolution time.
- Added -clobber switch to mhstore(1) to control overwriting of
  existing files.
- Added -outfile switch to mhstore(1).
- Added -noall/-all switches to sortm(1).  sortm -noall requires
  a messages argument.
- $PAGER overrides the compiled-in default pager command.
- Added etc/scan.highlighted format file, as an example of how to
  highlight/colorize the output of scan(1).
- inc(1) now supports a -port switch to specify the port used by the
  POP server.
- pick(1) now decodes MIME-encoded header fields before searching.
- The VISUAL and EDITOR environment variables are now supported as fallbacks
  if the user does not configure an editor entry in their profile.
- The format engine (mh_format(5)) now properly accounts for multibyte
  characters when accounting for column widths.
- burst(1) now can burst MIME-formatted digests (messages that contain
  message/rfc822 parts instead of messages formatted with RFC 934).
- All proc entries (showproc, moreproc, etc) can now accept entries that
  contain spaces and shell metacharacters.  If found, such entries will
  either be space-split or processed by /bin/sh.
- A new program, fmttest(1) is included to help debug format files.  See
  NOTABLE CHANGES below.
- mhshow/mhstore now have support for RFC-2017 (access-type=url) for
  external message bodies.
- Added -retainsequences switch to refile(1).
- A new program, mhfixmsg(1), is included to rewrite MIME messages with
  various transformations.  See NOTABLE CHANGES below.
- Added -[no]rmmproc switches to rmm(1).
- Added support for Content-Disposition header (RFC 2183) to mhstore(1)
  and mhn(1) when used with -auto.
- All nmh commands now support transactional locking for sequence files.
  See NOTABLE CHANGES below.
- There is no longer a per-folder maximum number of sequences.
- For the SMTP MTA TLS can now be negotiated at the beginning of the
  connection with the -initialtls switch.
- Messages can now be selected using a relative offset from some other
  message, or from the start or end of a sequence.  See mh-sequence(5).
- The -changecur and -nochangecur switches have been added to mhlist(1).
- mhbuild(1) can now encode 8-bit message headers using RFC-2047 encoding
  rules.  See NOTABLE CHANGES below.
- The whatnow(1) attach feature will determine the content type of an
  attachment using a program such as file --mime-type, if available at
  configuration time.  If not, it will use mhshow-suffix- entries as
  before.  The -v switch to attach causes it to display the mhbuild
  directive that send(1) will use.
- mhbuild(1) now supports the -auto/-noauto flags (to be used by send(1)
  when invoking mhbuild automatically).
- mhbuild(1) now is automatically run by send, to insure that all outgoing
  messages have proper MIME formatting.  See NOTABLE CHANGES below.
- A new header, "Attach", is supported by mhbuild; it is used to replace
  previous functionality (which by default used a header named
  Nmh-Attachment).  See NOTABLE CHANGES below.
- The default Content-Transfer-Encoding for text parts is now 8bit.
- mhbuild(1) now supports a selectable Content-Transfer-Encoding
- If nmh was configured with iconv(3) support, mhshow will convert, if
  necessary, the charset of text/plain content to match the user's
  locale setting.  See NOTABLE CHANGES below.
- Added support for %{charset} display string escape to mhshow(1).
- The MIME parsing and generating routines now support RFC 2231 extended
  parameter information.
- mh-mime(7) now provides an introduction to nmh's MIME handling.
- mhshow(1) will now by default display all text content under one pager,
  and display markers for non-text and non-inline content.  The content
  markers are changeable via mh-format(5).  See NOTABLE CHANGES below.

-----------------
OBSOLETE FEATURES
-----------------
- Changed exit status of each nmh command's -version and -help
  switches from 1 to 0.
- The following environment variables were deprecated in nmh 1.5
  and removed from this release:
    MHPOPDEBUG (use -snoop command line switch instead)
    MM_NOASK (use -nolist and -nopause command line switches instead)
    NOMHNPROC (use -nocheckmime command line switch instead)
    FACEPROC (undocumented faceproc feature removed)
- Changed repl and dist default to -noatfile.  The default of -atfile

uip/rmf.c
/* rmf.c -- remove a folder
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/utils.h>

#define RMF_SWITCHES \
    X("interactive", 0, INTRSW) \
    X("nointeractive", 0, NINTRSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(RMF);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(RMF, switches);
#undef X

/*
 * static prototypes
 */
static int rmf(char *);
static void rma (char *);


int
main (int argc, char **argv)
{
    int defolder = 0, interactive = -1;
    char *cp, *folder = NULL, newfolder[BUFSIZ];
    char buf[BUFSIZ], **argp, **arguments;

    if (nmh_init(argv[0], 1)) { return 1; }

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [+folder] [switches]",
			invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case INTRSW: 
		    interactive = 1;
		    continue;
		case NINTRSW: 
		    interactive = 0;
		    continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = pluspath (cp);
	} else {
	    adios (NULL, "usage: %s [+folder] [switches]", invo_name);
	}
    }

    if (!context_find ("path"))
	free (path ("./", TFOLDER));
    if (!folder) {
	folder = getfolder (1);
	defolder++;
    }
    if (strcmp (m_mailpath (folder), pwd ()) == 0)
	adios (NULL, "sorry, you can't remove the current working directory");

    if (interactive == -1)
	interactive = defolder;

    if (strchr (folder, '/') && (*folder != '/') && (*folder != '.')) {
	for (cp = stpcpy(newfolder, folder); cp > newfolder && *cp != '/'; cp--)
	    continue;
	if (cp > newfolder)
	    *cp = '\0';
	else
	    strncpy (newfolder, getfolder(0), sizeof(newfolder));
    } else {
	strncpy (newfolder, getfolder(0), sizeof(newfolder));
    }

    if (interactive) {
	cp = concat ("Remove folder \"", folder, "\"? ", NULL);
	if (!read_yes_or_no_if_tty (cp))
	    done (0);
	free (cp);
    }

    if (rmf (folder) == OK) {
	char *cfolder = context_find(pfolder);
	if (cfolder && strcmp (cfolder, newfolder)) {
	    printf ("[+%s now current]\n", newfolder);
	    context_replace (pfolder, newfolder);	/* update current folder */
	}
    }
    context_save ();	/* save the context file */
    done (0);
    return 1;
}

static int
rmf (char *folder)
{
    int i, others;
    char *maildir;
    char cur[BUFSIZ];
    struct dirent *dp;
    DIR *dd;

    switch (i = chdir (maildir = m_maildir (folder))) {
	case OK: 
	    if (access (".", W_OK) != NOTOK && access ("..", W_OK) != NOTOK)
		break;
	    /* FALLTHRU */

	case NOTOK: 
	    snprintf (cur, sizeof(cur), "atr-%s-%s",
			current, m_mailpath (folder));
	    if (!context_del (cur)) {
		printf ("[+%s de-referenced]\n", folder);
		return OK;
	    }
	    inform("you have no profile entry for the %s folder +%s",
		    i == NOTOK ? "unreadable" : "read-only", folder);
	    return NOTOK;
    }

    if ((dd = opendir (".")) == NULL)
	adios (NULL, "unable to read folder +%s", folder);
    others = 0;

    /*
     *	Run the external delete hook program.
     */

    (void)ext_hook("del-hook", maildir, NULL);

    while ((dp = readdir (dd))) {
	switch (dp->d_name[0]) {
	    case '.': 
		if (strcmp (dp->d_name, ".") == 0
			|| strcmp (dp->d_name, "..") == 0)
		    continue;
		/* FALLTHRU */

	    case ',': 
		break;

	    default: 
		if (m_atoi (dp->d_name))
		    break;
		if (strcmp (dp->d_name, LINK) == 0
			|| has_prefix(dp->d_name, BACKUP_PREFIX))
		    break;

		inform("file \"%s/%s\" not deleted, continuing...",
			folder, dp->d_name);
		others++;
		continue;
	}
	if (m_unlink (dp->d_name) == NOTOK) {
	    admonish (dp->d_name, "unable to unlink %s:", folder);
	    others++;
	}
    }

    closedir (dd);

    /*
     * Remove any relevant private sequences
     * or attributes from context file.
     */
    rma (folder);

    if (chdir ("..") < 0) {
	advise ("..", "chdir");
    }
    if (others == 0 && remdir (maildir))
	return OK;

    inform("folder +%s not removed", folder);
    return NOTOK;
}


/*
 * Remove all the (private) sequence information for
 * this folder from the profile/context list.
 */

static void
rma (char *folder)
{
    int alen, j, plen;
    char *cp;
    struct node *np, *pp;

    alen = LEN("atr-");
    plen = strlen (cp = m_mailpath (folder)) + 1;

    /*
     * Search context list for keys that look like
     * "atr-something-folderpath", and remove them.
     */
    for (np = m_defs, pp = NULL; np; np = np->n_next) {
	if (ssequal ("atr-", np->n_name)
		&& (j = strlen (np->n_name) - plen) > alen
		&& *(np->n_name + j) == '-'
		&& strcmp (cp, np->n_name + j + 1) == 0) {
	    if (!np->n_context)
		inform("bug: context_del(key=\"%s\"), continuing...",
		    np->n_name);
	    if (pp) {
		pp->n_next = np->n_next;
		np = pp;
	    } else {
		m_defs = np->n_next;
	    }
	    ctxflags |= CTXMOD;
	} else {
	    pp = np;
	}
    }
}
